File: Btech.Sql.Console/Models/Responses/Base/Response.cs
using Btech.Sql.Console.Extensions;
using Newtonsoft.Json;

namespace Btech.Sql.Console.Models.Responses.Base;

public class Response
{
    #region Public Constants

    public const string ErrorMessageJsonPropertyName = "errorMessage";
    public const string ErrorMessagesJsonPropertyName = "errorMessages";
    public const string ValidationErrorMessagesJsonPropertyName = "validationErrorMessages";

    #endregion Public Constants

    #region Public Properties

    [JsonProperty(ErrorMessageJsonPropertyName, NullValueHandling = NullValueHandling.Ignore)]
    public string ErrorMessage { get; set; }

    [JsonProperty(ErrorMessagesJsonPropertyName, NullValueHandling = NullValueHandling.Ignore)]
    public List<string> ErrorMessages { get; set; }

    [JsonProperty(ValidationErrorMessagesJsonPropertyName, NullValueHandling = NullValueHandling.Ignore)]
    public Dictionary<string, string> ValidationErrorMessages { get; set; }

    [JsonIgnore]
    public bool IsErrored => !this.ErrorMessage.IsNullOrEmpty() ||
                             this.ErrorMessages?.Any() is true ||
                             this.ValidationErrorMessages?.Any() is true;

    #endregion Public Properties
}

public class Response<T> : Response
{
    public const string DataJsonPropertyName = "data";

    [JsonProperty(DataJsonPropertyName, NullValueHandling = NullValueHandling.Ignore)]
    public T Data { get; set; }
}

File: Btech.Sql.Console/ClientApp/src/app/_models/responses/base/response.ts
export class ResponseBase {
    private _errorMessage?: string;
    private _errorMessages?: string[];
    private _validationErrorMessages?: { key: string, value: string }[];

    get errorMessage(): string | undefined {
        return this._errorMessage;
    }

    set errorMessage(value: string | undefined) {
        this._errorMessage = value;
    }

    get errorMessages(): string[] | undefined {
        return this._errorMessages;
    }

    set errorMessages(value: string[] | undefined) {
        this._errorMessages = value;
    }

    get validationErrorMessages(): { key: string, value: string }[] | undefined {
        return this._validationErrorMessages;
    }

    set validationErrorMessages(value: { key: string, value: string }[] | undefined) {
        this._validationErrorMessages = value;
    }
}

export class Response<T> extends ResponseBase {
    private _data?: T;

    get data(): T | undefined {
        return this._data;
    }

    set data(value: T | undefined) {
        this._data = value;
    }
}